Update work-pools.md (#9752)
PrefectHQ/prefect: docs/concepts/work-pools.md

Together work queue priority and concurrency enable precise control over work. For example, a pool may have three queues: A "low" queue with priority `10` and no concurrency limit, a "high" queue with priority `5` and a concurrency limit of `3`, and a "critical" queue with priority `1` and a concurrency limit of `1`. This arrangement would enable a pattern in which there are two levels of priority, "high" and "low" for regularly scheduled flow runs, with the remaining "critical" queue for unplanned, urgent work, such as a backfill.

> Priority is evaluated if more is scheduled than can be submitted. This can be due to Concurrency limits that limit the number of flows running, or could be due to infrastructure limits such as Kubernetes where there is insufficient CPU and/or Memory availability. In these scenarios where work is in a `SCHEDULED` state, priority becomes relevant. 

Reviewer: ```suggestion
Priority is only evaluated if more runs are scheduled than can be submitted. This can be due to concurrency limits that limit the number of flows running or infrastructure limits such as insufficient CPU and/or memory availability on Kubernetes.
```

I don't think we actually do anything to handle CPU/memory availability though. We'll just create the jobs and Kubernetes will schedule them as needed.

Reviewer: Also, priority still determines the _order_ in which runs are submitted even if they can all be submitted — right?

Author: For sure! I'll make those changes

> Priority for flow-run submission (transitioning from `SCHEDULED` to `PENDING`) proceeds from the highest priority to the lowest priority.  In the preceding example, all work from the "critical" queue (priority 1) will be submitted, before any work is submitted from "high" (priority 5). Once all work has been submitted from priority queue "critical", work from the "high" queue will begin submission. 

Author: Reworded to account for this - hopefully better?

> What happens if new work is received on the "critical" queue while work is still in scheduled on the "high" and "low" queues? We go back to ensuring all scheduled work is first satisfied from the highest priority queue, until it is empty in waterfall fashion.

Reviewer: I don't think we should write docs with question prompts like this. I'd just say it: "If new work is received ..."

Author: Updated!

Priority is evaluated to determine the order of flow-runs that can be submitted for execution. 
If all flow runs are capable of being executed with no limitation due to concurrency or otherwise, priority is still used to determine order of submission, but there is no impact to execution.
If not all flow runs can be executed, as a result of concurrency limits or infrastructure limitations, priority is used to determine which queues receive precedence to submit runs for execution, based upon the defined priority of that queue.

Priority for flow-run submission proceeds from the highest priority to the lowest priority.  In the preceding example, all work from the "critical" queue (priority 1) will be submitted, before any work is submitted from "high" (priority 5). Once all work has been submitted from priority queue "critical", work from the "high" queue will begin submission. 

> If new work is received on the "critical" queue while work is still in scheduled on the "high" and "low" queues, flow-run submission goes back to ensuring all scheduled work is first satisfied from the highest priority queue, until it is empty in waterfall fashion.

Reviewer 2: ```suggestion
If new flow runs are received on the "critical" queue while flow runs are still in scheduled on the "high" and "low" queues, flow run submission goes back to ensuring all scheduled work is first satisfied from the highest priority queue, until it is empty, in waterfall fashion.
```
